src/stream.rs

```diff
@@ -332,15 +389,23 @@ pub(crate) fn connect_host(unit: &Unit, hostname: &str, port: u16) -> Result<Tcp
 
     // rust's absurd api returns Err if we set 0.
     // Setting it to None will disable the native system timeout
-    if unit.timeout_read > 0 {
+    if let Some(deadline) = deadline {
+        stream
+            .set_read_timeout(Some(deadline - Instant::now()))
+            .ok();
+    } else if unit.timeout_read > 0 {
```

Comment:
I think it might be worth noting in the doc for the deprecated `read_timeout` that if one sets both that and `timeout`, then timeout takes precedence.